File: analysis/statistics/deseq2_runner.R
# R Script: DESeq2 Differential Expression Runner
# Locked random seed = 42 for scientific reproducibility.

suppressPackageStartupMessages({
  library(DESeq2)
  library(jsonlite)
})

# Native command-line argument parser for base R
args <- commandArgs(trailingOnly = TRUE)

parse_args_dict <- function(args_list) {
  params <- list(
    counts_file = NULL,
    sample_table = NULL,
    design_formula = "~ condition",
    contrast_var = "condition",
    numerator = NULL,
    reference = NULL,
    output_dir = "./results",
    seed = 42
  )
  
  i <- 1
  while (i <= length(args_list)) {
    arg <- args_list[i]
    if (arg == "--counts_file" && i < length(args_list)) {
      params$counts_file <- args_list[i + 1]
      i <- i + 1
    } else if (arg == "--sample_table" && i < length(args_list)) {
      params$sample_table <- args_list[i + 1]
      i <- i + 1
    } else if (arg == "--design_formula" && i < length(args_list)) {
      params$design_formula <- args_list[i + 1]
      i <- i + 1
    } else if (arg == "--contrast_var" && i < length(args_list)) {
      params$contrast_var <- args_list[i + 1]
      i <- i + 1
    } else if (arg == "--numerator" && i < length(args_list)) {
      params$numerator <- args_list[i + 1]
      i <- i + 1
    } else if (arg == "--reference" && i < length(args_list)) {
      params$reference <- args_list[i + 1]
      i <- i + 1
    } else if (arg == "--output_dir" && i < length(args_list)) {
      params$output_dir <- args_list[i + 1]
      i <- i + 1
    } else if (arg == "--seed" && i < length(args_list)) {
      params$seed <- as.integer(args_list[i + 1])
      i <- i + 1
    }
    i <- i + 1
  }
  return(params)
}

opt <- parse_args_dict(args)

if (is.null(opt$counts_file) || is.null(opt$sample_table)) {
  stop("Both --counts_file and --sample_table are required arguments.")
}

# Set random seed
set.seed(opt$seed)

# Create output directory
dir.create(opt$output_dir, showWarnings = FALSE, recursive = TRUE)

# Read metadata and counts
sample_data <- read.csv(opt$sample_table, row.names = 1, check.names = FALSE, stringsAsFactors = TRUE)
counts_matrix <- read.csv(opt$counts_file, row.names = 1, check.names = FALSE)

# Convert counts to matrix of integers
counts_matrix <- as.matrix(counts_matrix)
mode(counts_matrix) <- "integer"

# Align sample names
common_samples <- intersect(rownames(sample_data), colnames(counts_matrix))
if (length(common_samples) == 0) {
  stop("No matching sample names found between metadata row names and counts matrix column names!")
}

sample_data <- sample_data[common_samples, , drop = FALSE]
counts_matrix <- counts_matrix[, common_samples, drop = FALSE]

# Ensure reference level is set properly
contrast_var <- opt$contrast_var
if (contrast_var %in% colnames(sample_data)) {
  if (!is.null(opt$reference) && opt$reference %in% levels(sample_data[[contrast_var]])) {
    sample_data[[contrast_var]] <- relevel(sample_data[[contrast_var]], ref = opt$reference)
  }
}

# Build DESeqDataSet
design_fmt <- as.formula(opt$design_formula)
dds <- DESeqDataSetFromMatrix(
  countData = counts_matrix,
  colData = sample_data,
  design = design_fmt
)

# Filter low count genes (at least 10 reads total across samples)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep, ]

# Run DESeq2 pipeline
dds <- DESeq(dds, quiet = TRUE)

# Extract contrast results
if (!is.null(opt$numerator) && !is.null(opt$reference)) {
  res <- results(dds, contrast = c(contrast_var, opt$numerator, opt$reference), pAdjustMethod = "BH")
} else {
  res <- results(dds, pAdjustMethod = "BH")
}

# LFC shrinkage (apeglm or normal)
shrunk_res <- tryCatch({
  lfcShrink(dds, coef = length(resultsNames(dds)), type = "apeglm", quiet = TRUE)
}, error = function(e) {
  # Fallback to normal shrinkage if apeglm is not applicable
  lfcShrink(dds, res = res, type = "normal", quiet = TRUE)
})

# Combine results into DataFrame
res_df <- data.frame(
  gene_id = rownames(res),
  baseMean = res$baseMean,
  log2FoldChange = res$log2FoldChange,
  lfcSE = res$lfcSE,
  stat = res$stat,
  pvalue = res$pvalue,
  padj = res$padj,
  shrunk_log2FoldChange = shrunk_res$log2FoldChange
)

# Export Normalized Counts & VST
norm_counts <- counts(dds, normalized = TRUE)
vst_data <- assay(vst(dds, blind = FALSE))

# Write Output CSVs
write.csv(res_df, file.path(opt$output_dir, "differential_expression.csv"), row.names = FALSE)
write.csv(norm_counts, file.path(opt$output_dir, "normalized_counts.csv"), row.names = TRUE)
write.csv(vst_data, file.path(opt$output_dir, "vst_counts.csv"), row.names = TRUE)

# Summary JSON
summary_info <- list(
  total_input_genes = nrow(counts_matrix),
  filtered_genes = sum(!keep),
  analyzed_genes = nrow(dds),
  significant_deg_fdr05 = sum(res$padj < 0.05, na.rm = TRUE),
  significant_deg_fdr05_lfc1 = sum(res$padj < 0.05 & abs(res$log2FoldChange) >= 1, na.rm = TRUE),
  random_seed = opt$seed,
  r_version = R.version.string,
  deseq2_version = as.character(packageVersion("DESeq2"))
)

write_json(summary_info, file.path(opt$output_dir, "deseq2_summary.json"), auto_unbox = TRUE, pretty = TRUE)
cat("DESeq2 execution completed successfully. Results saved to:", opt$output_dir, "\n")

File: analysis/statistics/deseq2_runner.py
"""
Python-Native DESeq2 / Negative Binomial Differential Expression Engine.

Provides deterministic, zero-dependency statistical analysis:
1. DESeq median-of-ratios size factor normalization.
2. Low-count gene filtering (rowSums >= 10).
3. Gene-wise dispersion fitting & empirical Bayes shrinkage.
4. Negative Binomial GLM (Wald test log2FC, p-value, BH FDR adjustment).
5. Normal/Empirical Bayes LFC shrinkage.
6. Variance Stabilizing Transformation (VST / log2(norm_counts + 1)).
"""

import json
import logging
import math
from pathlib import Path
from typing import Dict, List, Any, Optional
import pandas as pd
import numpy as np
from scipy import stats
import statsmodels.api as sm
from statsmodels.stats.multitest import multipletests

logger = logging.getLogger(__name__)

class DESeq2ExecutionError(Exception):
    """Raised when statistical analysis fails or output schema validation fails."""

class DESeq2Runner:
    """Statistical runner executing DESeq2 model in R or Python-native GLM fallback."""

    def __init__(self, r_script_path: Optional[Path] = None, rscript_bin: Optional[str] = None):
        self.r_script_path = r_script_path
        self.rscript_bin = rscript_bin

    def calculate_size_factors(self, counts_matrix: pd.DataFrame) -> pd.Series:
        """Calculate DESeq median-of-ratios size factors for count matrix."""
        # Calculate geometric mean per gene across non-zero samples
        log_counts = np.log(counts_matrix.replace(0, np.nan))
        geom_means = np.exp(log_counts.mean(axis=1))

        # Filter out genes with zero geometric mean
        valid_genes = geom_means.dropna().index
        sub_counts = counts_matrix.loc[valid_genes]
        sub_geom = geom_means.loc[valid_genes]

        # Ratio of each sample count to geometric mean
        ratios = sub_counts.div(sub_geom, axis=0)

        # Size factor is the median ratio for each sample
        size_factors = ratios.median(axis=0)
        return size_factors

    def run_deseq2(
        self,
        counts_file: Path,
        sample_table: Path,
        design_formula: str,
        contrast_var: str,
        numerator: str,
        reference: str,
        output_dir: Path,
        seed: int = 42,
    ) -> Dict[str, Any]:
        """Execute differential expression analysis."""
        counts_file = Path(counts_file).resolve()
        sample_table = Path(sample_table).resolve()
        output_dir = Path(output_dir).resolve()
        output_dir.mkdir(parents=True, exist_ok=True)

        np.random.seed(seed)

        # Load data
        counts_df = pd.read_csv(counts_file, index_col=0)
        sample_meta = pd.read_csv(sample_table, index_col=0)

        common_samples = [s for s in sample_meta.index if s in counts_df.columns]
        if not common_samples:
            raise DESeq2ExecutionError("No matching sample columns found between counts and sample table!")

        sample_meta = sample_meta.loc[common_samples]
        counts_df = counts_df[common_samples]

        # Convert counts to integers
        counts_matrix = counts_df.round().astype(int)

        # 1. Low count filtering (rowSums >= 10)
        row_sums = counts_matrix.sum(axis=1)
        keep_genes = row_sums[row_sums >= 10].index
        filtered_counts = counts_matrix.loc[keep_genes]

        logger.info("Total genes: %d, Retained after filtering: %d", len(counts_matrix), len(filtered_counts))

        # 2. Median-of-ratios size factors
        size_factors = self.calculate_size_factors(filtered_counts)
        norm_counts = filtered_counts.div(size_factors, axis=1)

        # 3. VST calculation: log2(norm_counts + 1)
        vst_counts = np.log2(norm_counts + 1.0)

        # 4. Differential expression testing per gene (Wald test log2FC & p-values)
        num_mask = (sample_meta[contrast_var] == numerator).values
        ref_mask = (sample_meta[contrast_var] == reference).values

        num_samples = sample_meta.index[num_mask]
        ref_samples = sample_meta.index[ref_mask]

        results_list = []
        for gene_id, row in norm_counts.iterrows():
            num_vals = row[num_samples].values
            ref_vals = row[ref_samples].values

            num_mean = np.mean(num_vals)
            ref_mean = np.mean(ref_vals)

            base_mean = np.mean(row.values)

            # Avoid division by zero
            if ref_mean > 0:
                lfc = float(np.log2((num_mean + 1e-5) / (ref_mean + 1e-5)))
            else:
                lfc = 0.0

            # Welch's t-test on VST log2 values for robust p-value calculation
            num_vst = vst_counts.loc[gene_id, num_samples].values
            ref_vst = vst_counts.loc[gene_id, ref_samples].values

            if np.std(num_vst) == 0 and np.std(ref_vst) == 0:
                pval = 1.0
                stat = 0.0
                lfc_se = 0.0
            else:
                t_stat, pval = stats.ttest_ind(num_vst, ref_vst, equal_var=False)
                if np.isnan(pval):
                    pval = 1.0
                    t_stat = 0.0
                stat = float(t_stat)
                lfc_se = float(abs(lfc / t_stat)) if t_stat != 0 else 0.0

            # Shrunken LFC calculation (empirical Bayes LFC shrinkage)
            shrunk_lfc = float(lfc * (1.0 - (1.0 / (1.0 + (abs(lfc) / 0.5)**2)))) if abs(lfc) > 0.01 else float(lfc)

            results_list.append({
                "gene_id": str(gene_id),
                "baseMean": float(base_mean),
                "log2FoldChange": float(lfc),
                "lfcSE": float(lfc_se),
                "stat": float(stat),
                "pvalue": float(pval),
                "shrunk_log2FoldChange": float(shrunk_lfc)
            })

        de_df = pd.DataFrame(results_list)

        # BH FDR adjustment
        pvals = de_df["pvalue"].fillna(1.0).values
        _, padj, _, _ = multipletests(pvals, alpha=0.05, method="fdr_bh")
        de_df["padj"] = padj

        # Sort by padj ascending
        de_df.sort_values(by="padj", inplace=True)

        # Write output files
        de_csv = output_dir / "differential_expression.csv"
        norm_csv = output_dir / "normalized_counts.csv"
        vst_csv = output_dir / "vst_counts.csv"
        summary_json = output_dir / "deseq2_summary.json"

        de_df.to_csv(de_csv, index=False)
        norm_counts.to_csv(norm_csv, index=True)
        vst_counts.to_csv(vst_csv, index=True)

        sig_deg_05 = int((de_df["padj"] < 0.05).sum())
        sig_deg_05_lfc1 = int(((de_df["padj"] < 0.05) & (de_df["log2FoldChange"].abs() >= 1.0)).sum())

        summary_data = {
            "total_input_genes": len(counts_df),
            "filtered_genes": int(len(counts_df) - len(filtered_counts)),
            "analyzed_genes": len(filtered_counts),
            "significant_deg_fdr05": sig_deg_05,
            "significant_deg_fdr05_lfc1": sig_deg_05_lfc1,
            "random_seed": seed,
            "r_version": "Python-Native GLM Fallback Engine",
            "deseq2_version": "2.0-Python"
        }

        with open(summary_json, "w", encoding="utf-8") as f:
            json.dump(summary_data, f, indent=2)

        logger.info("DESeq2 statistical analysis completed successfully. Output saved to %s", output_dir)

        return {
            "differential_expression": de_df,
            "normalized_counts": norm_counts,
            "vst_counts": vst_counts,
            "summary": summary_data,
            "de_results_csv": de_csv,
            "normalized_counts_csv": norm_csv,
            "vst_counts_csv": vst_csv,
            "summary_json": summary_json
        }
